whisk/action_test.go

```diff
@@ -0,0 +1,151 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk
+
+import (
+    "testing"
+    "github.com/stretchr/testify/assert"
+    "net/http"
+    "io/ioutil"
+    "encoding/json"
+    "strings"
+    "net/url"
+)
+
+const (
+    NODE_ACTION_NO_CODE = `{
+        "name": "test",
+        "publish": false,
+        "annotations": [
+            {
+                "key": "exec",
+                "value": "nodejs:6"
+            }
+        ],
+        "version": "0.0.1",
+        "exec": {
+            "kind": "nodejs:6",
+            "binary": false
+        },
+        "parameters": [],
+        "limits": {
+            "timeout": 60000,
+            "memory": 256,
+            "logs": 10
+        },
+        "namespace": "test@openwhisk"
+    }`
+
+    NODE_ACTION = `{
+        "name": "test",
+        "publish": false,
+        "annotations": [
+            {
+                "key": "exec",
+                "value": "nodejs:6"
+            }
+        ],
+        "version": "0.0.1",
+        "exec": {
+            "kind": "nodejs:6",
+            "code": "...",
+            "binary": false
+        },
+        "parameters": [],
+        "limits": {
+            "timeout": 60000,
+            "memory": 256,
+            "logs": 10
+        },
+        "namespace": "test@openwhisk"
+    }`
+)
+
+type ActionResponse struct {
+    Body    string
+}
+
+type ActionRequest struct {
+    Method  string
+    URL     string
+}
+
+var actionResponse = &ActionResponse{}
+var actionRequest = &ActionRequest{}
+
+type MockClient struct {}
+
+func (c *MockClient) NewRequestUrl(method string, urlRelResource *url.URL, body interface{}, includeNamespaceInUrl bool, appendOpenWhiskPath bool, encodeBodyAs string, useAuthentication bool) (*http.Request, error) {
+    return &http.Request{}, nil
+}
+
+func (c *MockClient) NewRequest(method, urlStr string, body interface{}, includeNamespaceInUrl bool) (*http.Request, error) {
+    actionRequest.Method = method
+    actionRequest.URL = urlStr
+
+    request, err := http.NewRequest(method, urlStr, nil)
+    if (err != nil) {
+        return &http.Request{}, err
+    }
+
+    return request, nil
+}
+
+func (c *MockClient) Do(req *http.Request, v interface{}, ExitWithErrorOnTimeout bool, secretToObfuscate ...ObfuscateSet) (*http.Response, error) {
+    var reader = strings.NewReader(actionResponse.Body)
+
+    dc := json.NewDecoder(reader)
+    dc.UseNumber()
+    err := dc.Decode(v)
+
+    if err != nil {
+        return nil, err
+    }
+
+    resp := &http.Response{
+        StatusCode: 200,
+        Body: ioutil.NopCloser(reader),
+    }
+
+    return resp, nil
+}
+
+func TestActionGet(t *testing.T) {
+    assert := assert.New(t)
+    mockClient := &MockClient{}
+    actionService := &ActionService{client: mockClient}
+
+    actionResponse.Body = NODE_ACTION_NO_CODE
```

Comment:
instead of using globals to drive the test behavior, how about using the action name parameter instead.  pass in `testNoCode` as the action name in one test, and `testWithDummyCode` as the other action name.  in the mocked `Do()` method, check the action name to determine which mock response to return.

Reply:
Can't define a map as a const.

Reply:
not sure i understand.  what map would required to be a const?

Reply:
How would we map `testWithDummyCode` with `NODE_ACTION_NO_CODE` in a simple manner without a map?

Reply:
yeah, a map can be used, but i'm thinking that it does not have to be a const.